library/std/src/sys/hermit/thread_parker.rs

```diff
@@ -0,0 +1,117 @@
+use super::abi;
+use crate::pin::Pin;
+use crate::sync::atomic::AtomicU64;
+use crate::sync::atomic::Ordering::{Acquire, Relaxed, Release};
+use crate::time::{Duration, Instant};
+
+// These values are larger than u32::MAX, so they never conflict with the thread's `pid`.
+const EMPTY: u64 = 0x1_0000_0000;
+const NOTIFIED: u64 = 0x2_0000_0000;
+
+// Note about atomic memory orderings:
+// Acquire ordering is necessary to obtain the token, as otherwise the parked thread
+// could perform memory operations visible before it was unparked. Since once set,
+// the token cannot be unset by other threads, the state can be reset with a relaxed
+// store once it has been read with acquire ordering.
```

Comment:
I am not sure this is convincing to me. My understanding (perhaps flawed, I'm definitely not an expert in this) is that relaxed doesn't guarantee that other threads will observe the new empty state; we need a release store that synchronizes with the other thread's acquire ordering in order to guarantee correct behavior here.

In general, I'm not sure that hermit should deviate from the pattern of the general unix thread parker implementation, which uses SeqCst ordering for all of these operations.

Reply:
> My understanding (perhaps flawed, I'm definitely not an expert in this) is that relaxed doesn't guarantee that other threads will observe the new empty state; we need a release store that synchronizes with the other thread's acquire ordering in order to guarantee correct behavior here.

That's not correct. SeqCst or Release+Acquire also don't guarantee any other thread will directly observe the new value either. The specification only talks about ordering, not about timing. And even Relaxed guarantees a perfectly consistent ordering for all operations on the same atomic variable. Using stronger memory ordering only means that you can under some conditions guarantee that something that happened to _another_ piece of memory is observable once an atomic store/change is observed, but is unrelated to when that change is observed in the first place.